### ISSUE TITLE
Issues with resolving dependency with python 3.11.1

### ISSUE THREAD

**Issue by ketan-deserve** [pdm-project/pdm#1556]
Getting below error while running 
`pdm sync --clean --prod --no-self --verbose` 

ERROR: 
`resolvelib.resolvers.ResolutionImpossible: [RequirementInformation(requirement=FileRequirement(name=None, marker=None, extras=None, specifier=None, editable=False, prerelease=False, url=None, path=PosixPath('Django>=3.*'), subdirectory=None)]`

StackTrace : 

```
File "/usr/local/bin/pdm", line 8, in <module>
--
922 | sys.exit(main())
923 | ^^^^^^
924 | File "/usr/local/lib/python3.11/site-packages/pdm/core.py", line 258, in main
925 | return Core().main(args)
926 | ^^^^^^^^^^^^^^^^^
927 | File "/usr/local/lib/python3.11/site-packages/pdm/core.py", line 190, in main
928 | raise cast(Exception, err).with_traceback(traceback)
929 | File "/usr/local/lib/python3.11/site-packages/pdm/core.py", line 185, in main
930 | f(project, options)
931 | File "/usr/local/lib/python3.11/site-packages/pdm/cli/commands/sync.py", line 36, in handle
932 | actions.do_sync(
933 | File "/usr/local/lib/python3.11/site-packages/pdm/cli/actions.py", line 197, in do_sync
934 | candidates = resolve_candidates_from_lockfile(project, requirements)
935 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
936 | File "/usr/local/lib/python3.11/site-packages/pdm/cli/actions.py", line 140, in resolve_candidates_from_lockfile
937 | mapping, *_ = resolve(
938 | ^^^^^^^^
939 | File "/usr/local/lib/python3.11/site-packages/pdm/resolver/core.py", line 35, in resolve
940 | result = resolver.resolve(requirements, max_rounds)
941 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
942 | File "/usr/local/lib/python3.11/site-packages/resolvelib/resolvers.py", line 521, in resolve
943 | state = resolution.resolve(requirements, max_rounds=max_rounds)
944 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
945 | File "/usr/local/lib/python3.11/site-packages/resolvelib/resolvers.py", line 414, in resolve
946 | raise ResolutionImpossible(self.state.backtrack_causes)


```

python version : 3.11.1.
PDM version : 2.3.1

Dependencies: 
```
dependencies = [
    "Django~=3.2.*",
    "djangorestframework~=3.13.*",
    "pyjwt>=2.4.0",
    "pyjwt-rsa==0.3.9",
    "python-json-logger>=2.0.1",
    "pydantic>=1.10.2",
    "fastapi",
    "pytz",
    "boto3",
    "requests",
    "newrelic",
]
```

pdm install --no-self -v is giving output :

```
PDM 2.3.1 is installed, while 2.3.2 is available.
Please run `pdm self update` to upgrade.
Run `pdm config check_update false` to disable the check.
Lock file hash doesn't match pyproject.toml, packages may be outdated
Updating the lock file...
pdm.termui: ======== Start resolving requirements ========
pdm.termui:   b''
pdm.termui:   pyjwt>=2.4.0
pdm.termui:   pyjwt-rsa==0.3.9
pdm.termui:   python-json-logger>=2.0.1
pdm.termui:   pydantic>=1.10.2
pdm.termui:   fastapi
pdm.termui:   pytz
pdm.termui:   boto3
pdm.termui:   requests
pdm.termui:   newrelic
pdm.termui:   pytest
pdm.termui:   pytest-djangoapp
pdm.termui:   pytest-cov
pdm.termui:   moto[sns,sqs]
pdm.termui:   freezegun
pdm.termui:   requests-mock
pdm.termui:   pysqslistener
pdm.termui:   black
pdm.termui:   python>=3.7
pdm.termui:   Adding requirement b''
Traceback (most recent call last):
  File "/workspace/platform-custom-middleware/venv/bin/pdm", line 8, in <module>
    sys.exit(main())
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/core.py", line 258, in main
    return Core().main(args)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/core.py", line 190, in main
    raise cast(Exception, err).with_traceback(traceback)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/core.py", line 185, in main
    f(project, options)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/cli/commands/install.py", line 58, in handle
    actions.do_lock(
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/cli/actions.py", line 95, in do_lock
    mapping, dependencies = resolve(
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/resolver/core.py", line 35, in resolve
    result = resolver.resolve(requirements, max_rounds)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 521, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 372, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/resolvelib/resolvers.py", line 172, in _add_to_criteria
    if not criterion.candidates:
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/resolvelib/structs.py", line 127, in __bool__
    next(iter(self))
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/resolver/providers.py", line 261, in matches_gen
    yield from super_find()
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/resolver/providers.py", line 150, in matches_gen
    candidates = self._find_candidates(reqs[0])
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/resolver/providers.py", line 127, in _find_candidates
    can = make_candidate(requirement)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/models/candidates.py", line 623, in make_candidate
    return Candidate(req, name, version, link)
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/models/candidates.py", line 142, in __init__
    link = req.as_file_link()  # type: ignore
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/pdm/models/requirements.py", line 324, in as_file_link
    return Link(url)
  File "<string>", line 7, in __init__
  File "/workspace/platform-custom-middleware/venv/lib/python3.10/site-packages/unearth/link.py", line 48, in __post_init__
    if self.url.startswith(vcs_prefixes):
TypeError: a bytes-like object is required, not 'str'
```

**Comment by frostming** [comment]
Can you show the `pyproject.toml`, or preferrably the project directory?

**Comment by dashsatish** [comment]
@frostming this is how the pyproject.toml looks like:
```
[project]
name = "my-middleware"
version = "2.7.5"
description = "xxxxxxx"
authors = [
    { name = "....", email = "" },
]
dependencies = [
    "Django>=3.0.14",
    "djangorestframework>=3.13.1",
    "pyjwt>=2.4.0",
    "pyjwt-rsa==0.3.9",
    "python-json-logger>=2.0.1",
    "pydantic>=1.10.2",
    "fastapi",
    "pytz",
    "boto3",
    "requests",
    "newrelic",
]
requires-python = ">=3.7"
license = { text = "Proprietory" }

[build-system]
requires = ["pdm-pep517>=1.0.0", "setuptools>=42", "wheel", "setuptools_git_versioning"]
build-backend = "pdm.pep517.api"


[tool]
[tool.pdm]
[[tool.pdm.source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[tool.pdm.dev-dependencies]
dev = [
    "pytest",
    "pytest-djangoapp",
    "pytest-cov",
    "moto[sns,sqs]",
    "freezegun",
    "requests-mock",
    "pysqslistener",
    "black",
    "twine",
]
```

**Comment by frostming** [comment]
Closed by [2.3.3](https://github.com/pdm-project/pdm/releases/tag/2.3.3)
